`eslint-typegen` types are not compatible with ESLint types

<!--
  ❗Please don't ignore this template.❗

  If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
  The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
  Before posting the issue, please confirm that the problem you're getting
  is not related with your code editor configuration.
  To make sure it's not, run: yarn eslint src/your-file.vue
-->

**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.
<!-- If you do not read the FAQ and open an issue that is listed in the FAQ, we may silently close the issue. -->

**Tell us about your environment**

- **ESLint version:** 9.31.0
- **eslint-plugin-vue version:** 10.3.0
- **Vue version:** -
- **Node version:** 20.19.0 
- **Operating System:** Linux

**Please show your full configuration:**
(see reproduction repository - no special settings, just plain strict-mode typescript)

**What did you do?**	
<!-- Please include the actual source code causing the issue. -->
Updated `eslint-plugin-vue`

**What did you expect to happen?**
Everything to continue working, and for there to be no type errors

**What actually happened?**
<!--
  Please include the actual, raw output from ESLint.
  If you are only looking at the results of your editor extension, also check the CLI results.
-->

Lots of type errors

```
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/array-bracket-newline'' of type 'RuleEntry<VueArrayBracketNewline> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/array-bracket-spacing'' of type 'RuleEntry<VueArrayBracketSpacing> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/array-element-newline'' of type 'RuleEntry<VueArrayElementNewline> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/arrow-spacing'' of type 'RuleEntry<VueArrowSpacing> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/attribute-hyphenation'' of type 'RuleEntry<VueAttributeHyphenation> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/attributes-order'' of type 'RuleEntry<VueAttributesOrder> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/block-lang'' of type 'RuleEntry<VueBlockLang> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/block-order'' of type 'RuleEntry<VueBlockOrder> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/block-spacing'' of type 'RuleEntry<VueBlockSpacing> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/block-tag-newline'' of type 'RuleEntry<VueBlockTagNewline> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/brace-style'' of type 'RuleEntry<VueBraceStyle> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/camelcase'' of type 'RuleEntry<VueCamelcase> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
Error: node_modules/eslint/lib/types/index.d.ts(1374,3): error TS2411: Property ''vue/comma-dangle'' of type 'RuleEntry<VueCommaDangle> | undefined' is not assignable to 'string' index type 'RuleEntry<any[]>'.
...
```

The issue is because the original type (`RulesRecord`) defines an index meaning you cannot have optional properties - removing the `?` from the properties resolves the issue.

See [this playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3nAMsAdjFKOAXzgDMoIQ4ByFAZwBs8YaBYAKE4Hpu4ATFAGNGAQyhoQEfgFdGaOkxY0MPPnDi5RIBmFFC0OfIVUdeGiy0Jl9aAEpyGd4dH5wUADwK5+9OA-kAeTAYYAhcP3Q1CxiNGgA3GRRucShRRABaACM0oQBrFBgM3BQAd2YSmgB+AC5sKygAOgCUAFF8KEQAHlFcRAA+aJjiIZJoka4zPgabA39HYNDwyNG4xOTU9OzcgqKS8rwUarqjAiaW9phOnr7BqbHJ8wzn57h1lKg0zJz9XeKyipoF7PThIVBwABqSQAgp90gAhHaFAByAMOcAAvHAANoAXQAPtiABREgBEokYpXS9FJcHxcFJJXihFp9NJQmWwHo3hgpIAlHTTBoQHJQoDanAshAIPJepxhXgAJIEED0CVE3AyEBZEz0zWMRh8zjEPm46LAjJvJIfL7bX6FDL0PRCPAAczgFtByDQUJQsK+iPtMAAys63ZicQTieTKdTWQymSzTYSyRSqYgaYLGShmVB+QAaIVwehu+QQilJCVSmUoOUcDQQLIAK2EMHoitw-upVelstw8rgmwzHa7GZ7NbrxDNkxmtjgAFlEC16M4OVB+ABGIsabFQRx1blQN24k4NZqOS6dADcxs4nFnc0Xy9Xri3nm8vgXS8cK5c68WYQRBgEz3sYUCzGgT4-i+-4hIBKz1nA3AAFTIQOcDIXA7RkNAcyAjktZ5H4ohkGccAQKguDhr0bg6jhEhwCIEAlrg7pDpKSJtnAeBwAABl0KpgGIBD9Lx6GYQAAvQKBoAAFjAMBgPQNS8AwFQwI06xNvQjTQK63B7vI9A2lsPz5A6+yAo08kgIw4ncAOCTWkOdrmXsaKVKeYHnvIl7dL6o6Bm5qIHCUdwaChaGIRhWG4PRcwchEXI8sWYasdxSWCIOcKIBxQZ+Dx-GCcJKCieJcBSTJcDyYpymqUo+CaUk2m6VA+mGQwJnfJxjppa61kwLZ9mOe8LlmX8Tr6G6xz1N5FwdP5MI5UFuyhlNrF3CBHAPpB35GTB-AAEzbjiHUHlcx5eWcPltAtN4cFtO1fs+f5HUWkXldhuFoPhEiiERg6kSYFEoFR6U0ZKKD0WgTEsWxOV5W5BW4HxAkoOAJVldFknSXJClKSp3BqSwTUoC1ekGT+XWuX8lmHANQ3Yw5iFORsOU0xZHlHFdhA3X5XQBctnEhYC4VIahn1xd9jGctyoPwJNLrpXgJZZex42FMjqPFaIIlidjFW49V+N1UTDUaVpOkUx1xljT1itugzdlMyNznsxrRQO6xM2nLz81XItfpC0Ga1K66m1AA) showcasing the issue both when merging directly into the interface and when using `extends` (which results in the error being put on the original interface, masking it somewhat).

Note that in the playground you cannot reproduce the issue using ESLint itself due to how packages work in the playground.

**Repository to reproduce this issue**
<!--
  ❗ Please share a repository that can reproduce your issue.
  ❗ If you don't share it, we just label it as requesting a repository that can reproduce your issue.
  ❗ If you don't share it for a few days, we silently close this issue.
-->
https://github.com/G-Rath/bug-eslint-vue

1. clone repo
2. run `npm ci`
3. run `npm run typescript`

Alternatively, look at the GitHub Actions runs

--- a/.changeset/eight-camels-refuse.md
+++ b/.changeset/eight-camels-refuse.md
@@ -1,0 +1,5 @@
+---
+"eslint-plugin-vue": patch
+---
+
+Resolved TypeScript compatibility issues introduced by eslint-typegen

--- a/tools/generate-typegen.mjs
+++ b/tools/generate-typegen.mjs
@@ -2,8 +2,13 @@ import fs from 'node:fs/promises'
 import { pluginsToRulesDTS } from 'eslint-typegen/core'
 import plugin from '../lib/index.js'
 
-const dts = await pluginsToRulesDTS({
-  vue: plugin
-})
+const dts = await pluginsToRulesDTS(
+  {
+    vue: plugin
+  },
+  {
+    includeAugmentation: false
+  }
+)
 
 await fs.writeFile('lib/eslint-typegen.d.ts', dts)

